docs/_posts/2021-03-29-using-seahorn.md

```diff
@@ -0,0 +1,331 @@
+
+## Compiling Rust for verification
+
+The [Rust compiler works][MIR blog post]
+in four stages:
+first, it compiles Rust source code to HIR (the high-level IR);
+HIR is converted to MIR (the mid-level IR);
+MIR is converted to LLVM IR (the low-level virtual machine IR);
+and, finally, it compiles LLVM IR down to machine code and links the result.
+
+When using an LLVM-based tool like [SeaHorn], we need to modify this behaviour
+so that the LLVM IR (instead of machine code) is linked and saved to a file.
+This can be done by passing extra flags to `rustc`, instructing it to link the
+bitcode files.
+When using `cargo` these flags are passed through the `RUSTFLAGS` environment
+variable.
+
+```
+export RUSTFLAGS="-Clto -Cembed-bitcode=yes --emit=llvm-bc $RUSTFLAGS"
+```
+
+We also have to pass some configuration flags to `cargo` and `rustc` to
+configure the `verification-annotations` crate correctly.
+
+```
+export RUSTFLAGS="--cfg=verify $RUSTFLAGS"
+```
+
+Our goal is to find bugs so we turn on some additional error checking
+in the compiled code to detect arithmetic overflows.
+
+```
+export RUSTFLAGS="-Warithmetic-overflow -Coverflow-checks=yes $RUSTFLAGS"
+```
+
+And, when we find a bug, we want to report it as efficiently as
+possible so we make sure that the program will abort if it panics.
+
+
+```
+export RUSTFLAGS="-Zpanic_abort_tests -Cpanic=abort $RUSTFLAGS"
+```
+
+With all those definitions, we can now run
+
+```
+cargo build --features=verifier-seahorn
+```
+
+Depending on which platform you are running on, the resulting LLVM bitcode file
+may be placed placed in `target/debug/deps/try_seahorn.bc` (OSX) or in a file with
+a name like `target/debug/deps/try_seahorn-*.bc` (Linux).
+At least for simple cases, we can refer to both files as
+`target/debug/deps/try_seahorn*.bc` and not worry about the exact filename.
+
+
+<!-- ### Compiling large programs -->
+
+<!-- Finally, on larger, more complex projects than this example, we have seen -->
+<!-- problems with the Rust compiler generating SSE instructions that KLEE does not -->
+<!-- support. -->
+<!-- We don't have a complete solution for this but we have found that it helps -->
+<!-- to compile with a low (but non-zero) level of optimization, to disable -->
+<!-- more sophisticated optimizations and to disable SSE and AVX features. -->
+
+<!-- ``` -->
+<!-- export RUSTFLAGS="-Copt-level=1 $RUSTFLAGS" -->
+<!-- export RUSTFLAGS="-Cno-vectorize-loops -Cno-vectorize-slp $RUSTFLAGS" -->
+<!-- export RUSTFLAGS="-Ctarget-feature=-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2 $RUSTFLAGS" -->
+<!-- ``` -->
+
+<!-- (This is not needed for our simple example.) -->
+
+
+## Running SeaHorn
+
+Having built a bitcode file containing the program and all the libraries that it
+depends on, we now need to fix a few things in the bitcode file to make it
+suitable for [SeaHorn].
+This includes deleting the `main` function rustc added, replacing panic handling
```

Comment:
I think it would be worth describing each of these in more detail - if they care about the other details about how to build LLVM bitcode and run seahorn, then they probably want this detail too.

Reply:
In particular, the most interesting bit is "what is the problem" (eg the code in print is far too complex for SeaHorn to handle or whatever).